hedera-node/hedera-app/src/main/java/com/hedera/node/app/workflows/handle/FinalizeContextImpl.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright (C) 2023 Hedera Hashgraph, LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.node.app.workflows.handle;
+
+import static java.util.Objects.requireNonNull;
+
+import com.hedera.hapi.node.base.AccountID;
+import com.hedera.node.app.service.token.TokenService;
+import com.hedera.node.app.service.token.records.FinalizeContext;
+import com.hedera.node.app.workflows.dispatcher.ReadableStoreFactory;
+import com.hedera.node.app.workflows.dispatcher.WritableStoreFactory;
+import com.hedera.node.app.workflows.handle.record.SingleTransactionRecordBuilderImpl;
+import com.hedera.node.app.workflows.handle.stack.Savepoint;
+import com.hedera.node.app.workflows.handle.stack.SavepointStackImpl;
+import com.swirlds.config.api.Configuration;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.time.Instant;
+
+/**
+ * The default implementation of {@link FinalizeContext}.
+ */
+public class FinalizeContextImpl implements FinalizeContext {
+    private final AccountID payer;
+    private final SingleTransactionRecordBuilderImpl recordBuilder;
+    private final SavepointStackImpl stack;
+    private final WritableStoreFactory writableStoreFactory;
+    private ReadableStoreFactory readableStoreFactory;
+
+    /**
+     * Constructs a {@link FinalizeContextImpl}.
+     *
+     * @param payer              The {@link AccountID} of the payer
+     * @param recordBuilder      The main {@link SingleTransactionRecordBuilderImpl}
+     * @param stack              The {@link SavepointStackImpl} used to manage savepoints
+     */
+    public FinalizeContextImpl(
+            @NonNull final AccountID payer,
+            @NonNull final SingleTransactionRecordBuilderImpl recordBuilder,
+            @NonNull final SavepointStackImpl stack) {
+        this.payer = requireNonNull(payer, "payer must not be null");
+        this.recordBuilder = requireNonNull(recordBuilder, "recordBuilder must not be null");
+        this.stack = requireNonNull(stack, "stack must not be null");
+        this.writableStoreFactory = new WritableStoreFactory(stack, TokenService.NAME);
```

Comment:
Rather than creating a writable store factory here that only supports `TokenService`, should we be creating it when needed to support the requested WritableStore type, rather than only supporting `TokenService`?

Reply:
Finalize only exists for the token service. It is defined in the token service API and is only used there.